File: Programmers/level2/주차_요금_계산/주차_요금_계산.py
import math


def solution(fees, records):
    in_ = dict()
    fee = dict()

    for x in records:
        a, b, c = x.split()
        h, m = a.split(':')
        time = int(h) * 60 + int(m)

        if c == 'IN':
            in_[b] = time
        else:
            fee[b] = fee.get(b, 0) + (time - in_[b])  # dict.get(key, 없는 경우 반환 값)
            del in_[b]

    for x in in_:  # out이 없는 경우
        fee[x] = fee.get(x, 0) + 23 * 60 + 59 - in_[x]

    answer = []
    for x, y in sorted(fee.items()):
        if y > fees[0]:
            answer.append(fees[1] + math.ceil((y - fees[0]) / fees[2]) * fees[3])
        else:
            answer.append(fees[1])

    return answer

File: Programmers/level2/주차_요금_계산/주차_요금_계산.java
import java.util.*;

class Solution {
    public int[] solution(int[] fees, String[] records) {

        Map<String, Integer> m = new TreeMap<String, Integer>(); // key 이름에 따라 정렬
        Map<String, Integer> input = new HashMap<String, Integer>();

        for (String r : records) {
            String[] v = r.split(" "); // t, car, state
            String[] t = v[0].split(":");
            int time = Integer.parseInt(t[0])*60 + Integer.parseInt(t[1]);

            if (v[2].equals("IN")) { // .equals()
                input.put(v[1], time); // map.put()
            } else {
                int tm = time - input.get(v[1]);
                if (m.containsKey(v[1])) { // 출차 기록
                    m.put(v[1], m.get(v[1]) + tm);
                } else {
                    m.put(v[1], tm);
                }
                input.remove(v[1]);
            }
        }

        // 수정: 아직 출차하지 않은 차량 처리
        for(String c : input.keySet()) {
            int time = 23 * 60 + 59 - input.get(c);
            if (m.containsKey(c)) {
                m.put(c, m.get(c) + time);
            } else {
                m.put(c, time);
            }
        }

        int[] answer = new int[m.size()];
        int i = 0;

        for (String c : m.keySet()) {
            int tm = fees[1];
            int rest = m.get(c) - fees[0];
            if (rest > 0) {
                if (rest % fees[2] != 0) {
                    tm += fees[3];
                }
                tm += fees[3] * (rest / fees[2]);
            }
            answer[i] = tm;
            i++;
        }

        // Arrays.sort(answer); 값 정렬 방법
        return answer;
    }
}
